tests/e2e/collection/grpc.go

```diff
@@ -0,0 +1,551 @@
+			"not own NFT",
+			fmt.Sprintf("%s/lbm/collection/v1/contracts/%s/balances/%s/%s", val.APIAddress, s.contractID, s.stranger.String(), tokenID),
+			false,
+			&collection.QueryBalanceResponse{},
+			&collection.QueryBalanceResponse{
+				Balance: collection.Coin{TokenId: tokenID, Amount: cmath.ZeroInt()},
+			},
+		},
+		{
+			"invalid contract ID",
+			fmt.Sprintf("%s/lbm/collection/v1/contracts/%s/balances/%s/%s", val.APIAddress, "wrong id", s.customer.String(), tokenID),
+			true,
+			&collection.QueryBalanceResponse{},
+			&collection.QueryBalanceResponse{},
+		},
+		{
+			"invalid token ID",
+			fmt.Sprintf("%s/lbm/collection/v1/contracts/%s/balances/%s/%s", val.APIAddress, s.contractID, s.customer.String(), "wrong id"),
+			true,
+			&collection.QueryBalanceResponse{},
+			&collection.QueryBalanceResponse{},
+		},
+		{
+			"invalid address",
+			fmt.Sprintf("%s/lbm/collection/v1/contracts/%s/balances/%s/%s", val.APIAddress, s.contractID, "wrong address", tokenID),
+			true,
+			&collection.QueryBalanceResponse{},
+			&collection.QueryBalanceResponse{},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		s.Run(tc.name, func() {
+			resp, err := testutil.GetRequest(tc.url)
+			s.Require().NoError(err)
+
+			err = s.cfg.Codec.UnmarshalJSON(resp, tc.respType)
+
+			if tc.expErr {
+				s.Require().Error(err)
+			} else {
+				s.Require().NoError(err)
+				s.Require().Equal(tc.expectResp.String(), tc.respType.String())
+			}
+		})
+	}
+}
+
+func (s *E2ETestSuite) TestBalancesGRPC() {
+	val := s.network.Validators[0]
+	tokenID := s.mintNFT(s.contractID, s.vendor, s.customer, s.nftClassID)
+
+	testCases := []struct {
+		name     string
+		url      string
+		expErr   bool
+		respType proto.Message
+	}{
+		{
+			"valid request",
+			fmt.Sprintf("%s/lbm/collection/v1/contracts/%s/balances/%s", val.APIAddress, s.contractID, s.vendor.String()),
+			false,
+			&collection.QueryAllBalancesResponse{},
+		},
+		{
+			"invalid contract ID",
+			fmt.Sprintf("%s/lbm/collection/v1/contracts/%s/balances/%s/%s", val.APIAddress, "wrong id", s.vendor.String(), tokenID),
+			true,
+			&collection.QueryAllBalancesResponse{},
+		},
+		{
+			"invalid address",
+			fmt.Sprintf("%s/lbm/collection/v1/contracts/%s/balances/%s/%s", val.APIAddress, s.contractID, "wrong address", tokenID),
+			true,
+			&collection.QueryAllBalancesResponse{},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		s.Run(tc.name, func() {
+			resp, err := testutil.GetRequest(tc.url)
+			s.Require().NoError(err)
+
+			err = s.cfg.Codec.UnmarshalJSON(resp, tc.respType)
+
+			if tc.expErr {
+				s.Require().Error(err)
+			} else {
+				s.Require().NoError(err)
+				s.Require().GreaterOrEqual(len(tc.respType.(*collection.QueryAllBalancesResponse).Balances), 1)
+			}
+		})
+	}
+}
+
+func (s *E2ETestSuite) TestStatisticsGRPC() {
+	val := s.network.Validators[0]
+	tokenID := s.mintNFT(s.contractID, s.vendor, s.vendor, s.nftClassID)
+	_ = s.burnNFT(s.contractID, s.vendor, tokenID)
+
```

Comment:
You may apply `introduce variable` for repeated URL formats.
Like this 
`
	const supplyUrlFormat = "%s/lbm/collection/v1/contracts/%s/token_types/%s/supply"
	const mintedUrlFormat = "%s/lbm/collection/v1/contracts/%s/token_types/%s/minted"
	const burntUrlFormat = "%s/lbm/collection/v1/contracts/%s/token_types/%s/burnt"
`